File: vane-core/src/main/java/org/oddlama/vane/core/material/HeadMaterialLibrary.java
package org.oddlama.vane.core.material;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.bukkit.NamespacedKey;
import org.json.JSONArray;

public class HeadMaterialLibrary {

    private static final Map<NamespacedKey, HeadMaterial> registry = new HashMap<>();
    private static final Map<String, List<HeadMaterial>> categories = new HashMap<>();
    private static final Map<String, List<HeadMaterial>> tags = new HashMap<>();
    private static final Map<String, HeadMaterial> byTexture = new HashMap<>();

    public static void load(final String string) {
        final var json = new JSONArray(string);
        for (int i = 0; i < json.length(); ++i) {
            // Deserialize
            final var mat = HeadMaterial.from(json.getJSONObject(i));

            // Add to registry
            registry.put(mat.key(), mat);
            byTexture.put(mat.texture(), mat);

            // Add to category lookup
            var category = categories.computeIfAbsent(mat.category(), k -> new ArrayList<>());
            category.add(mat);

            // Add to tag lookup
            for (final var tag : mat.tags()) {
                var tagList = tags.computeIfAbsent(tag, k -> new ArrayList<>());
                tagList.add(mat);
            }
        }
    }

    public static HeadMaterial from(final NamespacedKey key) {
        return registry.get(key);
    }

    public static HeadMaterial fromTexture(final String base64Texture) {
        return byTexture.get(base64Texture);
    }

    public static Collection<HeadMaterial> all() {
        return registry.values();
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/material/HeadMaterialLibrary.kt
package org.oddlama.vane.core.material

import org.bukkit.NamespacedKey
import org.json.JSONArray
import org.oddlama.vane.core.material.HeadMaterial.Companion.from

/**
 * Registry of known custom head materials loaded from JSON data.
 */
object HeadMaterialLibrary {
    /** Head registry by namespaced key. */
    private val registry   = mutableMapOf<NamespacedKey, HeadMaterial>()
    /** Category lookup map. */
    private val categories = mutableMapOf<String, MutableList<HeadMaterial>>()
    /** Tag lookup map. */
    private val tags       = mutableMapOf<String, MutableList<HeadMaterial>>()
    /** Lookup by base64 texture payload. */
    private val byTexture  = mutableMapOf<String, HeadMaterial>()

    /** Loads head materials from serialized JSON array content. */
    @JvmStatic
    fun load(string: String) {
        val json = JSONArray(string)
        for (i in 0 until json.length()) {
            val mat = from(json.getJSONObject(i))

            registry[mat.key!!] = mat
            byTexture[mat.texture!!] = mat

            categories.getOrPut(mat.category!!) { mutableListOf() }.add(mat)

            mat.tags.forEach { tag ->
                tags.getOrPut(tag!!) { mutableListOf() }.add(mat)
            }
        }
    }

    /** Resolves a head material by key. */
    @JvmStatic
    fun from(key: NamespacedKey?): HeadMaterial? = registry[key]

    /** Resolves a head material by texture payload. */
    @JvmStatic
    fun fromTexture(base64Texture: String?): HeadMaterial? = byTexture[base64Texture]

    /** Returns all loaded head materials. */
    @JvmStatic
    fun all(): Collection<HeadMaterial> = registry.values
}
